cost/aws/gp3_volume_upgrade/aws_upgrade_to_gp3_volume.pt

```diff
@@ -133,55 +152,55 @@ datasource "ds_unattached_volumes_list" do
           field "tagValue", xpath(col_item, "value")
         end
       end
-      field "attachments" do
-        collect xpath(col_item, "attachmentSet/item") do
-          field "instance_id", xpath(col_item, "instanceId")
-        end
-      end
       field "volume_id", xpath(col_item, "volumeId")
       field "availability_zone", xpath(col_item, "availabilityZone")
       field "snapshot_id", xpath(col_item, "snapshotId")
       field "volume_type", xpath(col_item, "volumeType")
       field "region", val(iter_item, "region")
       field "createdTime", xpath(col_item, "createTime")
       field "size", xpath(col_item, "size")
-      field "status", xpath(col_item,"status")
+      field "iops", xpath(col_item, "iops")
+      field "status", xpath(col_item, "status")
     end
   end
 end
 
-datasource "ds_unattached_volumes_map" do
-  run_script $js_filter_aws_unattached_volumes, $ds_unattached_volumes_list, $param_exclude_tags
+datasource "ds_superseded_volumes" do
+  run_script $js_filter_aws_superseded_volumes, $ds_superseded_volumes_list, $param_exclude_tags
 end
 
-datasource "ds_currency_reference" do
-  request do
-    host "raw.githubusercontent.com"
-    path "/rightscale/policy_templates/master/cost/scheduled_reports/currency_reference.json"
-    header "User-Agent", "RS Policies"
-  end
-end
-
-datasource "ds_flexera_optima" do
-  run_script $js_flexera_optima, $param_flexera_org_id_for_optima, rs_org_id
+datasource "ds_volume_types" do
+  run_script $js_volume_types, $ds_superseded_volumes
 end
 
-datasource "ds_currency_code" do
+datasource "ds_aws_products" do
```

Comment:
How big is this payload?

Reply:
When it iterates over all four volume types, here's the resource usage from `fpt`:

> ### Resource Usage
>   * StartedAt: Fri, 17 Dec 2021 22:38:09 UTC
>   * FinishedAt: Fri, 17 Dec 2021 22:38:14 UTC
>   * Duration: 5.635450443s
>   * Net IO:
>     * **datasource:ds_aws_products: 0.282 mb**
>     * datasource:ds_billing_centers: 0.169 mb
>     * datasource:ds_get_caller_identity: 0.000 mb
>     * datasource:ds_regions_list: 0.004 mb
>     * datasource:ds_superseded_volumes_list: 0.007 mb
>     * datasource:ds_volume_costs: 0.001 mb
>     * ds_aws_products: 0.000 mb
>     * ds_volume_costs: 0.003 mb
>   * Mem Usage (MB):
>     * **datasource:ds_aws_products: 0.044 mb**
>     * datasource:ds_billing_centers: 0.053 mb
>     * datasource:ds_flexera_optima: 0.000 mb
>     * datasource:ds_get_caller_identity: 0.000 mb
>     * datasource:ds_regions: 0.000 mb
>     * datasource:ds_regions_list: 0.001 mb
>     * datasource:ds_superseded_volumes: 0.002 mb
>     * datasource:ds_superseded_volumes_list: 0.002 mb
>     * datasource:ds_top_level_billing_centers: 0.018 mb
>     * datasource:ds_volume_cost_mapping: 0.002 mb
>     * datasource:ds_volume_costs: 0.001 mb
>     * datasource:ds_volume_types: 0.000 mb
>   * Durations (s):
>     * **datasource:ds_aws_products: 0.246 s** (4 calls)
>     * **datasource:ds_aws_products:net_io: 0.151 s** (4 calls)
>     * **datasource:ds_aws_products:parse: 0.002 s** (4 calls)
>     * datasource:ds_billing_centers: 0.257 s (1 call)
>     * datasource:ds_billing_centers:net_io: 0.206 s (1 call)
>     * datasource:ds_billing_centers:parse: 0.002 s (1 call)
>     * datasource:ds_get_caller_identity: 0.054 s (1 call)
>     * datasource:ds_get_caller_identity:net_io: 0.035 s (1 call)
>     * datasource:ds_get_caller_identity:parse: 0.000 s (1 call)
>     * datasource:ds_regions_list: 0.100 s (1 call)
>     * datasource:ds_regions_list:net_io: 0.076 s (1 call)
>     * datasource:ds_regions_list:parse: 0.000 s (1 call)
>     * datasource:ds_superseded_volumes_list: 3.719 s (8 calls)
>     * datasource:ds_superseded_volumes_list:net_io: 3.635 s (8 calls)
>     * datasource:ds_superseded_volumes_list:parse: 0.000 s (8 calls)
>     * datasource:ds_volume_costs: 1.125 s (1 call)
>     * datasource:ds_volume_costs:net_io: 0.714 s (1 call)
>     * datasource:ds_volume_costs:parse: 0.000 s (1 call)
>     * script:js_filter_aws_superseded_volumes: 0.006 s (1 call)
>     * script:js_flexera_optima: 0.017 s (1 call)
>     * script:js_get_costs: 0.006 s (1 call)
>     * script:js_regions: 0.006 s (1 call)
>     * script:js_top_level_bc: 0.009 s (1 call)
>     * script:js_volume_cost_mapping: 0.013 s (1 call)
>     * script:js_volume_types: 0.005 s (1 call)

Things are much faster when AWS lets you filter down to almost only what you want. 😉

Reply:
That is awesome, much better.